libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFSecureEncryptionKey.m

```diff
@@ -0,0 +1,193 @@
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFSecureEncryptionKey.h"
+#import "SFSDKCryptoUtils.h"
+#import "NSData+SFAdditions.h"
+
+// NSCoding constants
+static NSString * const kSecureEncryptionKeyCodingValue = @"com.salesforce.encryption.securekey.label";
+
+@interface SFSecureEncryptionKey () {
+    SecKeyRef publicKeyRef;
+    SecKeyRef privateKeyRef;
+}
+
+@property (nonatomic, strong) NSString *label;
+
+@end
+
+@implementation SFSecureEncryptionKey
+
+#pragma mark - Factory methods and constructor
+
++ (instancetype) createKey:(NSString*)label
+{
+    [SFSDKCryptoUtils deleteECKeyPairWithName:label]; // Delete existing key if any
+    return [[SFSecureEncryptionKey alloc] initWithLabel:label autoCreate:YES];
+}
+
++ (instancetype) retrieveKey:(NSString*)label
+{
+    return [[SFSecureEncryptionKey alloc] initWithLabel:label autoCreate:NO];
+}
+
++ (void) deleteKey:(NSString*)label
+{
+    [SFSDKCryptoUtils deleteECKeyPairWithName:label];
+}
+
+- (instancetype) initWithLabel:(NSString*)label autoCreate:(BOOL)autoCreate
+{
+    self = [super init];
+    if (self) {
+        self.label = label;
+        
+        // Use existing key pair if available
+        if ([self getKeyRefs]) {
+            return self;
+        }
+        
+        // Does not exist and should not be auto created
+        if (!autoCreate) {
+            return nil;
+        }
+        
+        // Does not exist and should be auto created
+        BOOL useEnclave = [SFSDKCryptoUtils isSecureEnclaveAvailable];
+        [SFSDKCryptoUtils createECKeyPairWithName:label
+                              accessibleAttribute:kSecAttrAccessibleAlways
+                                 useSecureEnclave:useEnclave];
+        
+        // Creation successful
+        if ([self getKeyRefs]) {
+            [SFSDKCoreLogger i:[self class] format:@"Creating secure key %@using secure enclave", useEnclave ? @"" : @"NOT "];
+            return self;
+        }
+        // Creation failed
+        else {
+            return nil;
+        }
+    }
+    return self;
+}
+
+- (void) dealloc {
```

Comment:
Is this file not ARC-enabled either?

Reply:
The compiler does not automatically manage the lifetimes of Core Foundation objects; you must call CFRetain and CFRelease.
https://developer.apple.com/library/archive/releasenotes/ObjectiveC/RN-TransitioningToARC/Introduction/Introduction.html#//apple_ref/doc/uid/TP40011226-CH1-SW1